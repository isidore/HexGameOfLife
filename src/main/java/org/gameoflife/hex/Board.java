package org.gameoflife.hex;


import com.spun.util.FormattedException;

import java.util.List;

public class Board {
    private List<Cell> liveCells;

    public static boolean isValidCoordinate2(Coordinates gridAt) {
        return isValidCoordinate(gridAt.getX(), gridAt.getY());
    }

    public List<Cell> getLiveCells() {
        return liveCells;
    }

    public void setLiveCells(List<Cell> liveCells) {
        this.liveCells = liveCells;
    }

    void setAlive(int x, int y) {
        if (!isValidCoordinate(x, y)) {
            throw new FormattedException("Invalid Location for (%s, %s)", x, y);
        }

        liveCells.add(new Cell(x, y));
    }

    public static boolean isValidCoordinate(int x, int y) {
        boolean xIsEven = x % 2 == 0;
        boolean yIsEven = y % 2 == 0;

        return yIsEven == xIsEven;
    }

    public boolean isAlive(Cell cell) {
        return liveCells.contains(cell);
    }

    @Override
    public String toString() {
        return HexPrinter.print(this);
    }

    public boolean isAlive(int x, int y) {
        return isAlive(new Cell(x, y));
    }
}
