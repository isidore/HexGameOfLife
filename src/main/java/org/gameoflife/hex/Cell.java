package org.gameoflife.hex;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Objects;

public class Cell {
    private int x;
    private int y;

    public Cell(int x, int y) {

        this.x = x;
        this.y = y;
    }

    public static Cell create(int x, int y) {
        return new Cell(x,y);
    }

    public static List<Cell> getLevelOneNeighbours(int x, int y) {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x - 1, y-1));
        cells.add(new Cell(x - 1, y+1));
        cells.add(new Cell(x , y-2));
        cells.add(new Cell(x , y+2));
        cells.add(new Cell(x +1, y-1));
        cells.add(new Cell(x + 1, y+1));

        return cells;
    }

    public static List<Cell> getLevelTwoNeighbours(int x, int y) {
        List<Cell> cells = new ArrayList<>();
        cells.add(new Cell(x-2, y));
        cells.add(new Cell(x -1, y+3));
        cells.add(new Cell(x - 1, y-3));
        cells.add(new Cell(x + 1, y+3));
        cells.add(new Cell(x +1, y-3));
        cells.add(new Cell(x+2 , y));

        return cells;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Cell cell = (Cell) o;
        return x == cell.x && y == cell.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }

    public List<Cell> getLevelOneNeighbours() {
        return getLevelOneNeighbours(x,y);
    }

    public List<Cell> getLevelTwoNeighbours() {
        return getLevelTwoNeighbours(x, y);
    }
}
