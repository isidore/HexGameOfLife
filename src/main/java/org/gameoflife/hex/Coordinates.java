package org.gameoflife.hex;

public class Coordinates {

    private int y;
    private int x;

    public Coordinates(int x, int y) {
        this.y = y;
        this.x = x;
    }

    public int getY() {
        return y;
    }

    public int getX() {
        return x;
    }
}
