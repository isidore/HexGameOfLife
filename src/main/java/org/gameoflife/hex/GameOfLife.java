package org.gameoflife.hex;

import org.apache.commons.lang.math.DoubleRange;
import org.lambda.query.Queryable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class GameOfLife {

    final Board board;

    public GameOfLife() {
        this(new ArrayList<>());
    }

    public GameOfLife(List<Cell> liveCells) {
        this.board = new Board();
        this.board.setLiveCells(liveCells);
    }

    static boolean isAlive2(List<Cell> liveCells, Cell cell) {
        return liveCells.contains(cell);
    }

    public GameOfLife advanceTurn() {
        Queryable<Cell> nextLivingCells = getLiveCellsAndNeighbours()
                .where(c -> survivesToNextTurn(getNeighbourScore(board, c), board.isAlive(c)));

        return new GameOfLife(nextLivingCells);
    }

    public static boolean survivesToNextTurn(double sum, boolean alive) {
        DoubleRange survivable = new DoubleRange(2, 3.3);
        DoubleRange growth = new DoubleRange(2.3, 2.9);

        boolean survives = survivable.containsDouble(sum) && alive;
        boolean born = growth.containsDouble(sum);
        return survives || born;
    }

    public static double getNeighbourScore(Board board, Cell cell) {
        double score = 0.0;
        score += getScore(board, cell.getLevelOneNeighbours(), 1);
        score += getScore(board, cell.getLevelTwoNeighbours(), 0.3);
        return score;
    }

    private static double getScore(Board board, List<Cell> neighbours, double weight) {
        double score = 0;

        for (Cell cell : neighbours) {
            if (board.isAlive(cell)) {
                score += weight;
            }
        }

        return score;
    }

    public void setAlive(int x, int y) {
        this.board.setAlive(x, y);
    }

    private Queryable<Cell> getLiveCellsAndNeighbours() {
        Set<Cell> liveCellsAndNeighbours = new HashSet<>(board.getLiveCells());

        for (Cell cell : board.getLiveCells()) {
            liveCellsAndNeighbours.addAll(cell.getAllNeighbours());
        }

        return Queryable.as(new ArrayList<>(liveCellsAndNeighbours));
    }

    @Override
    public String toString() {
        return this.board.toString();
    }

    public boolean isAlive(Cell cell) {
        return board.isAlive(cell);
    }

    public static boolean isValidCoordinates(Coordinates coordinates){
        return Board.isValidCoordinate(coordinates);
    }
}
