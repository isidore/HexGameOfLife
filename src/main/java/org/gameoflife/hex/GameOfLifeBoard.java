package org.gameoflife.hex;

import com.spun.util.FormattedException;

import java.util.*;

public class GameOfLifeBoard {
    private List<Cell> liveCells;

    public GameOfLifeBoard() {
        liveCells = new ArrayList<>();
    }

    public GameOfLifeBoard(List<Cell> liveCells) {

        this.liveCells = liveCells;
    }

    public static boolean isValidCoordinate(int x, int y) {
        boolean xIsEven = x%2 == 0;
        boolean yIsEven = y%2 == 0;

        return yIsEven == xIsEven;
    }

    public boolean isAlive(Cell cell) {
        return liveCells.contains(cell);
    }

    public void setAlive(int x, int y) {
        if (!isValidCoordinate(x, y)) {
            throw new FormattedException("Invalid Location for (%s, %s)", x, y);
        }

        liveCells.add(new Cell(x, y));
    }

    @Override
    public String toString() {
        return HexPrinter.print(this);
    }


    public GameOfLifeBoard advanceTurn() {
        ArrayList<Cell> nextLivingCells = new ArrayList<>();
        for (Cell cell : getLiveCellsAndNeighbours()) {
            if (isAliveNextTurn(getNeighbourScore(cell), isAlive(cell))) {
                nextLivingCells.add(cell);
            }
        }
        return new GameOfLifeBoard(nextLivingCells);
    }

    private static boolean isAliveNextTurn(double sum, boolean alive) {
        return 2 <= sum && sum <= 3.3 && alive;
    }

    private double getNeighbourScore(Cell cell) {
        return getScore(cell.getLevelOneNeighbours(), 1) + getScore(cell.getLevelTwoNeighbours(), 0.3);
    }

    private double getScore(List<Cell> neighbours, double weight) {
        double tempScore = 0;
        for (Cell c : neighbours) {
            if (isAlive(c)) {
                tempScore += weight;
            }
        }
        return tempScore;
    }


    private Set<Cell> getLiveCellsAndNeighbours() {
        Set<Cell> liveCellsAndNeighbours = new HashSet<>(liveCells);

        for (Cell cell : liveCells) {
            liveCellsAndNeighbours.addAll(cell.getAllNeighbours());
        }

        return liveCellsAndNeighbours;
    }

}
