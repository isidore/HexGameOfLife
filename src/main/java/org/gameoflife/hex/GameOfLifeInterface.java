package org.gameoflife.hex;

import org.apache.commons.lang.math.DoubleRange;

public interface GameOfLifeInterface {
    static boolean survivesToNextTurn(double sum, boolean alive) {
        DoubleRange survivable = new DoubleRange(2, 3.3);
        DoubleRange growth = new DoubleRange(2.3, 2.9);

        boolean survives = survivable.containsDouble(sum) && alive;
        boolean born = growth.containsDouble(sum);
        return survives || born;
    }

    static boolean isValidCoordinates(Coordinates coordinates) {
        return Board.isValidCoordinate(coordinates);
    }

    GameOfLife advanceTurn();

    void setAlive(int x, int y);

    @Override
    String toString();

    boolean isAlive(Cell cell);
}
