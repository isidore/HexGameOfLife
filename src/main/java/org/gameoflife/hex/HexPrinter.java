package org.gameoflife.hex;

import dk.ilios.asciihexgrid.AsciiBoard;
import dk.ilios.asciihexgrid.printers.SmallPointyAsciiHexPrinter;

public class HexPrinter {
    public static String print(GameOfLife board) {
        AsciiBoard asciiBoard = new AsciiBoard(0, 100, 0, 100, new SmallPointyAsciiHexPrinter());
        int size = 10;
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                if (GameOfLife.isValidCoordinates(new Coordinates(x, y))) {
                    if (board.getLiveCells().contains(new Cell(x, y))) {
                        asciiBoard.printHex("X", "X", 'X', translateX(x, y), translateY(x, y));
                    } else {
                        asciiBoard.printHex(x + "," + y, "", ' ', translateX(x, y), translateY(x, y));

                    }
                }
            }
        }
        return asciiBoard.prettPrint(true);
    }

    private static int translateY(int x, int y) {
        return y;
    }

    private static int translateX(int x, int y) {
        int emptyspaces = (x + 1) / 2;
        //return 10 + x - (y / 2) - emptyspaces;
        return x - emptyspaces;
    }
}
