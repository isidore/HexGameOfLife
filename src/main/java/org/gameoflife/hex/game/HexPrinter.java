package org.gameoflife.hex.game;

import dk.ilios.asciihexgrid.AsciiBoard;
import dk.ilios.asciihexgrid.printers.SmallPointyAsciiHexPrinter;

public class HexPrinter {
    public static String print(HexGameOfLife game) {
        AsciiBoard asciiBoard = new AsciiBoard(0, 100, 0, 100, new SmallPointyAsciiHexPrinter());
        int size = 10;
        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                if (HexGameOfLife.isValidCoordinates(new Coordinates(x, y))) {
                    if (game.isAlive(new Cell(x, y))) {
                        asciiBoard.printHex("X", "X", 'X', translateX(x, y), translateY(x, y));
                    } else {
                        asciiBoard.printHex(x + "," + y, "", ' ', translateX(x, y), translateY(x, y));
                    }
                }
            }
        }
        return asciiBoard.prettPrint(true);
    }

    private static int translateY(int x, int y) {
        return y;
    }

    private static int translateX(int x, int y) {
        int emptyspaces = (x + 1) / 2;
        return x - emptyspaces;
    }
}
