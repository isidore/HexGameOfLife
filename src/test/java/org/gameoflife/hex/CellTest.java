package org.gameoflife.hex;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

public class CellTest {

    @Test
    void getLevelOneNeighbours() {
        GameOfLife gameOfLife = GameOfLifeTest.createGameWithNeighbours(6, 0, true);
        Approvals.verify(HexPrinter.print(gameOfLife.board));
    }

    @Test
    void getLevelTwoNeighbours() {
        GameOfLife gameOfLife = GameOfLifeTest.createGameWithNeighbours(0, 6, true);
        Approvals.verify(HexPrinter.print(gameOfLife.board));
    }

    @Test
    void testGetAllNeighbours() {
        Cell cell = new Cell(4, 4);
        List<Cell> allNeighbours = new ArrayList<>(cell.getAllNeighbours());
        GameOfLife gameOfLife = new GameOfLife(allNeighbours);
        Approvals.verify(HexPrinter.print(gameOfLife.board));
    }
}
