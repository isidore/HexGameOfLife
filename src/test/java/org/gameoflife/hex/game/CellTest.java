package org.gameoflife.hex.game;

import org.approvaltests.Approvals;
import org.gameoflife.hex.HexPrinter;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

public class CellTest {

    @Test
    void getLevelOneNeighbours() {
        HexGameOfLife hexGameOfLife = HexGameOfLifeTest.createGameWithNeighbours(6, 0, true);
        Approvals.verify(HexPrinter.print(hexGameOfLife));
    }

    @Test
    void getLevelTwoNeighbours() {
        HexGameOfLife hexGameOfLife = HexGameOfLifeTest.createGameWithNeighbours(0, 6, true);
        Approvals.verify(HexPrinter.print(hexGameOfLife));
    }

    @Test
    void testGetAllNeighbours() {
        Cell cell = new Cell(4, 4);
        List<Cell> allNeighbours = new ArrayList<>(cell.getAllNeighbours());
        HexGameOfLife hexGameOfLife = new HexGameOfLife(allNeighbours);
        Approvals.verify(HexPrinter.print(hexGameOfLife));
    }
}
