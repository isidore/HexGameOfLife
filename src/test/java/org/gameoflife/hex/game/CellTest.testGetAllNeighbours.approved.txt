| = = = = = = = = = = = = = = = = = = = = = = = |
|    / \     / \     / \     / \     / \        |
|  /     \ /     \ /     \ /     \ /     \      |
| |  0,0  |  2,0  |  4,0  |  6,0  |  8,0  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,1  |  3,1  |  5,1  |  7,1  |  9,1  | |
|     |       |       |       |       |       | |
|    / \     / \     /X\     / \     / \     /  |
|  /     \ /     \ /X X X\ /     \ /     \ /    |
| |  0,2  |  2,2  |X  X  X|  6,2  |  8,2  |     |
| |       |       |X  X  X|       |       |     |
|  \     /X\     /X\X X X/X\     /X\     / \    |
|    \ /X X X\ /X X X\X/X X X\ /X X X\ /     \  |
|     |X  X  X|X  X  X|X  X  X|X  X  X|  9,3  | |
|     |X  X  X|X  X  X|X  X  X|X  X  X|       | |
|    / \X X X/X\X X X/ \X X X/X\X X X/ \     /  |
|  /     \X/X X X\X/     \X/X X X\X/     \ /    |
| |  0,4  |X  X  X|  4,4  |X  X  X|  8,4  |     |
| |       |X  X  X|       |X  X  X|       |     |
|  \     /X\X X X/X\     /X\X X X/X\     / \    |
|    \ /X X X\X/X X X\ /X X X\X/X X X\ /     \  |
|     |X  X  X|X  X  X|X  X  X|X  X  X|  9,5  | |
|     |X  X  X|X  X  X|X  X  X|X  X  X|       | |
|    / \X X X/ \X X X/X\X X X/ \X X X/ \     /  |
|  /     \X/     \X/X X X\X/     \X/     \ /    |
| |  0,6  |  2,6  |X  X  X|  6,6  |  8,6  |     |
| |       |       |X  X  X|       |       |     |
|  \     / \     / \X X X/ \     / \     / \    |
|    \ /     \ /     \X/     \ /     \ /     \  |
|     |  1,7  |  3,7  |  5,7  |  7,7  |  9,7  | |
|     |       |       |       |       |       | |
|    / \     / \     / \     / \     / \     /  |
|  /     \ /     \ /     \ /     \ /     \ /    |
| |  0,8  |  2,8  |  4,8  |  6,8  |  8,8  |     |
| |       |       |       |       |       |     |
|  \     / \     / \     / \     / \     / \    |
|    \ /     \ /     \ /     \ /     \ /     \  |
|     |  1,9  |  3,9  |  5,9  |  7,9  |  9,9  | |
|     |       |       |       |       |       | |
|      \     / \     / \     / \     / \     /  |
|        \ /     \ /     \ /     \ /     \ /    |
| = = = = = = = = = = = = = = = = = = = = = = = |
